docs/designs/015-custom-drain-extensibility.md

```diff
@@ -0,0 +1,237 @@
+# ADR-015: Behavior — Node Drain Extensibility
+
+## Context
+
+Current flow of node drainer essentially runs `kubectl drain` and when exactly a node will be drained in a cluster is configurable by configmap using `mode` which are: `Immediate`, `DeleteAfterTimeout` and `AllowCompletion`. These three modes essentially covers the entire lifecycle of workload. This works in most of the cases but there are some cases where additional coordination with scheduler (or meta scheduler) present in the cluster in required. Hence, we need a way to extend the functionality of node drain to make it aware of the ecosystem around scheduler present in the cluster.
+
+## Decision
+
+Design a node drain API using Custom Resource Definitions (CRDs) to represent custom drain operations. This API will be consumed by scheduler-specific bridge controllers to perform custom drain logic.
+
+## Implementation
+
+### Architecture
+
+```mermaid
+flowchart TD
+    A[Health Event] --> B[Node-Drainer Evaluator]
+    B --> C{Custom Drain Enabled?}
+    C -->|Yes| D[Load Template from ConfigMap]
+    C -->|No| E[Standard Drain Mode]
+    D --> F[Render Template with Event Data]
+    F --> G[Create Customer CR via Dynamic Client]
+    G --> H[Watch CR Status via Informer]
+    H --> I{Status Phase?}
+    I -->|Complete| J[Proceed with Pod Eviction]
+    I -->|Failed| K[Error/Retry]
+    I -->|Pending/InProgress| H
+```
+
+### Template Variables
+
+```go
+type TemplateData struct {
+  HealthEvent   HealthEevent
+  PodsToDrain   map[string][]string
+}
+```
+
+### Configuration (TOML)
+
+```toml
+[customDrain]
+    enabled = false
+    templateMountPath = "/etc/drain-templates"
+    templateFileName = "drain-template.yaml"
+    namespace = "nvsentinel"
+    apiGroup = "nvsentinel.nvidia.com"
+    version = "v1alpha1"
+    kind = "SlurmDrainRequest"
+
+    timeout = "3600s"
+
+    # Status watching (follows Kubernetes standard: InProgress, Complete, Failed)
+    statusConditionType = "Complete"
+    statusConditionStatus = "True"
+```
+
+### Template Example
+
+```yaml
+apiVersion: drain.example.com/v1alpha1
+kind: DrainRequest
+metadata:
+    name: drain-{{ .HealthEvent.NodeName }}-{{ .HealthEvent.EventID }}
```

Comment:
How do we enforce the name in the templates since this is user provided? Would we have some validation? Can you please add some info about that?